PurchasesCoreSwift/Purchasing/ProductInfo.swift

```diff
@@ -0,0 +1,156 @@
+                      normalDuration: String?,
+                      introDuration: String?,
+                      introDurationType: RCIntroDurationType,
+                      introPrice: NSDecimalNumber?,
+                      subscriptionGroup: String?,
+                      discounts: [PromotionalOffer]?) {
+        self.productIdentifier = productIdentifier
+        self.paymentMode = paymentMode
+        self.currencyCode = currencyCode
+        self.price = price
+        self.normalDuration = normalDuration
+        self.introDuration = introDuration
+        self.introDurationType = introDurationType
+        self.introPrice = introPrice
+        self.subscriptionGroup = subscriptionGroup
+        self.discounts = discounts
+    }
+
+    @objc public func asDictionary() -> [String: NSObject] {
+        var dict: [String: NSObject] = [:]
+        dict["product_id"] = productIdentifier as NSString
+        dict["price"] = price
+        dict["currency"] = currencyCode as NSObject
+
+        if paymentMode != .none {
+            dict["payment_mode"] = NSNumber(value: paymentMode.rawValue)
+        }
+
+        if let introPrice = introPrice {
+            dict["introductory_price"] = introPrice
+        }
+
+        if let subscriptionGroup = subscriptionGroup as NSString? {
+            dict["subscription_group_id"] = subscriptionGroup
+        }
+
+        if discounts != nil {
+            dict.merge(self.discountsAsDictionary()) { (_, new) in new }
+        }
+
+        dict.merge(self.productDurationsAsDictionary()) { (_, new) in new }
+
+        return dict
+    }
+
+    @objc public var cacheKey: String {
+        var key = """
+        \(productIdentifier)-\(price)-\(currencyCode)-\(paymentMode.rawValue)-\(introPrice ?? 0)-\
+        \(subscriptionGroup ?? "")-\(normalDuration ?? "")-\(introDuration ?? "")-\(introDurationType.rawValue)
+        """
+
+        guard let discounts = discounts else {
+            return key
+        }
+
+        if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
+            for offer in discounts {
+                key += "-\(offer.offerIdentifier ?? "null offer id")"
+            }
+        }
+        return key
+    }
+
+    private func discountsAsDictionary() -> [String: NSObject] {
+        var discountDict: [String: NSObject] = [:]
+            if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
+                if let discounts = self.discounts {
+                    let offers = NSMutableArray()
+                    for discount in discounts {
+                        guard let offerIdentifier = discount.offerIdentifier else {
+                            break
+                        }
+
+                        offers.add(["offer_identifier": offerIdentifier,
+                                    "price": discount.price,
+                                    "payment_mode": NSNumber(value: discount.paymentMode.rawValue)])
+
+                    }
```

Comment:
this block could be simplified as
```swift
let offers = discounts
    .filter { $0.offerIdentifier != nil }
    .map { discount in
        [
            "offer_identifier": discount.offerIdentifier,
            "price": discount.price,
            "payment_mode": NSNumber(value: discount.paymentMode.rawValue)
        ]
}
```

🤔 the offer identifier is nullable, so we might need unwrapping there, but that's the gist of it. 

sidenote: I don't think the NSNumber conversion here is necessary - it'll automatically get translated when bridging back to objc if needed

Reply:
The reason we can't do this is because on the ObjC side, we're veryyyyy loose with types when it comes to dictionaries. filter/map want stronger types, so that's cool, we can do:
`[String: [[String: NSObject]]]`
When we return from here, the caller is like, "wait, wut? I thought this was a `[String: NSObject]`"
So you keep going down this rabbit hole until you want to buy a farm in Scotland and raise sheep instead of battling the type system.

Reply:
re: `NSNumber` you're right 😄